daemon/access.go

```diff
@@ -151,3 +154,93 @@ func (ac snapAccess) CheckAccess(d *Daemon, r *http.Request, ucred *ucrednet, us
 	// FIXME: should snapctl access be allowed on the main socket?
 	return Forbidden("access denied")
 }
+
+var (
+	cgroupSnapNameFromPid = cgroup.SnapNameFromPid
+	requireThemeApiAccess = requireThemeApiAccessImpl
+)
+
+func requireThemeApiAccessImpl(d *Daemon, ucred *ucrednet) *apiError {
+	if ucred == nil {
+		return Forbidden("access denied")
+	}
+
+	switch ucred.Socket {
+	case dirs.SnapdSocket:
+		// Allow access on main snapd.socket
+		return nil
+
+	case dirs.SnapSocket:
+		// Handled below
+	default:
+		return Forbidden("access denied")
+	}
+
+	// Access on snapd-snap.socket requires a connected
+	// snap-themes-control plug.
+	snapName, err := cgroupSnapNameFromPid(int(ucred.Pid))
+	if err != nil {
+		return Forbidden("could not determine snap name for pid: %s", err)
+	}
+
+	st := d.state
+	st.Lock()
+	defer st.Unlock()
+	conns, err := ifacestate.ConnectionStates(st)
+	if err != nil {
+		return Forbidden("internal error: cannot get connections: %s", err)
+	}
+	for refStr, connState := range conns {
+		if connState.Undesired || connState.HotplugGone || connState.Interface != "snap-themes-control" {
+			continue
+		}
+		connRef, err := interfaces.ParseConnRef(refStr)
+		if err != nil {
+			return Forbidden("internal error: %s", err)
+		}
+		if connRef.PlugRef.Snap == snapName {
+			return nil
+		}
+	}
```

Comment:
Looks fine. We may want to grow a helper for this some day as this type of looping over conns is becoming more common (just a remark for me, nothing to do in this PR).